silence/silence.go

```diff
@@ -544,104 +540,179 @@ func validateSilence(s *pb.Silence) error {
+	existing, ok := s.getSilence(sil.Id)
+	// updateExisting is true iff this is an existing silence, and the update does
+	// not affect how alerts are matched or the time window in which the
+	// silence is active (excluding time extensions).
+	updateExisting := ok && canUpdate(existing, sil, now)
+	if !updateExisting {
+		// If this is a new silence, or it is replacing an existing silence,
+		// then give it an ID and make sure the start time is not in the past.
+		uid, err := uuid.NewV4()
+		if err != nil {
+			return fmt.Errorf("failed to create UUID: %w", err)
+		}
+		sil.Id = uid.String()
+		if sil.StartsAt.Before(now) {
+			sil.StartsAt = now
 		}
 	}
+	sil.UpdatedAt = now
 
-	msil := &pb.MeshSilence{
-		Silence:   sil,
-		ExpiresAt: sil.EndsAt.Add(s.retention),
-	}
-	b, err := marshalMeshSilence(msil)
-	if err != nil {
-		return err
+	// Check the limits before expiring the existing silence to prevent
+	// partial updates when limits are exceeded. Limits are checked against
+	// the mesh silence as this is the silence that is gossiped and stored
+	// on disk.
+	msil := s.toMeshSilence(sil)
+	if err := s.checkLimits(msil, updateExisting); err != nil {
+		return fmt.Errorf("exceeds limits: %w", err)
+	}
+
+	// expireExisting is true iff this is an existing silence, and the update
+	// affects how alerts are matched or the time window in which the silence
+	// is active (excluding extensions). If existing and sil reference the
+	// same memory, then the operation will fail because the new ID does not
+	// exist. This happens when the caller does not use Clone().
+	expireExisting := ok && !updateExisting && getState(existing, now) != types.SilenceStateExpired
+	if expireExisting {
+		if err := s.expire(existing.Id); err != nil {
+			return fmt.Errorf("failed to expire silence: %w", err)
+		}
 	}
 
-	// Check the limit unless the silence has been expired. This is to avoid
-	// situations where silences cannot be expired after the limit has been
-	// reduced.
+	return s.mergeAndBroadcast(msil, now)
+}
+
+func (s *Silences) checkLimits(msil *pb.MeshSilence, updateExisting bool) error {
 	if s.limits.MaxSilenceSizeBytes != nil {
 		n := msil.Size()
-		if m := s.limits.MaxSilenceSizeBytes(); m > 0 && n > m && sil.EndsAt.After(now) {
-			return fmt.Errorf("silence exceeded maximum size: %d bytes (limit: %d bytes)", n, m)
+		if m := s.limits.MaxSilenceSizeBytes(); m > 0 && n > m {
+			return fmt.Errorf("maximum size: %d bytes (limit: %d bytes)", n, m)
+		}
+	}
+	if s.limits.MaxSilences != nil && !updateExisting {
+		if m := s.limits.MaxSilences(); m > 0 && len(s.st)+1 > m {
+			return fmt.Errorf("maximum number of silences: %d (limit: %d)", len(s.st), m)
 		}
 	}
+	return nil
+}
 
+func (s *Silences) getSilence(id string) (*pb.Silence, bool) {
+	msil, ok := s.st[id]
+	if !ok {
+		return nil, false
+	}
+	return msil.Silence, true
+}
+
+func (s *Silences) mergeAndBroadcast(msil *pb.MeshSilence, now time.Time) error {
```

Comment:
This is the old [`setSilence`](https://github.com/prometheus/alertmanager/blob/v0.27.0/silence/silence.go#L550-L574) function. I removed validation and limits checks, and renamed it to `mergeAndBroadcast`.